api/submit_result.go

```diff
@@ -23,21 +27,49 @@ func newResultSubmitter(api *API) *resultSubmitter {
 }
 
 // Submit submits results for executionID.
-func (s *resultSubmitter) Submit(executionID string, outputKey string, outputData map[string]interface{}) error {
-	exec, err := s.api.execDB.Find(executionID)
+func (s *resultSubmitter) Submit(executionID string, outputKey string, outputData []byte) error {
+	exec, stateChanged, err := s.processExecution(executionID, outputKey, outputData)
+	if stateChanged {
+		// only publish to listeners when the execution's state changed.
+		go pubsub.Publish(exec.Service.ResultSubscriptionChannel(), exec)
+	}
+	// always return any error to the service.
+	return err
+}
+
+// processExecution processes execution and marks it as complated or failed.
+func (s *resultSubmitter) processExecution(executionID string, outputKey string, outputData []byte) (exec *execution.Execution, stateChanged bool, err error) {
+	stateChanged = false
+	exec, err = s.api.execDB.Find(executionID)
```

Comment:
Here is a race condition, there is no lock between find & save

Reply:
👍

Reply:
Fix in https://github.com/mesg-foundation/core/pull/652